### PR TITLE
Datastore refactor

### PR DESCRIPTION
**Pull request by curtisr7**
Hello -

I'd like a pluggable mechanism that will allow for me to switch the store backing sinopia. The current fs based approach is fine for most environments, but I would like the option to write my data elsewhere. database, cloud based, etc, etc, etc

This isn't yet a fully baked pull request, but I had to do a bit of re-plumbing so I wanted to be sure that @rlidwka is on board before going much further. (Note that I didn't actually implement a plugin.) If I'm feeling frisky one of these nights I might try to write a small aws plugin though.

Thoughts, feelings? I also wrote down a few of my [code review comments](https://github.com/curtisr7/sinopia/wiki/Refactor-thoughts) if you're interested.

### CODE REVIEW

**Comment by curtisr7** [comment]
@rlidwka any chance you have found some time to look through the direction of this refactor?

File: lib/local-storage.js [hunk 1]
```diff
@@ -484,158 +499,122 @@ Storage.prototype.get_package = function(name, options, callback) {
...
-        return callback(err)
-      }
-    }
+//
 
-    self._normalize_package(json)
-    updateFn(json, function(err) {
-      if (err) return callback(err)
+Storage.prototype.update_package = function(package, updateFn, _callback) {
+  // TODO -- is this function missing. I'm not certain how the previous code we wouldn't have storage.
```
Comment (rlidwka) [inline lib/local-storage.js:556]: `storage` can be specified in the config as `null` which means "don't store this package anywhere". A few people asked for this feature.

File: lib/local-storage.js [hunk 2]
```diff
@@ -21,6 +21,20 @@ function Storage(config) {
...
+  if(self.config.store == undefined) {
+    self.config.store = {'./disk-store': 'disk'}
+  }
+  var store = load_plugins(config, self.config.store, {
+    config : config,
+    logger : self.logger,
+  }, function(p) {
+    return true
+  })
+  // TODO -- support multiple stores? I vote no. Bomb out in this case
```
Comment (rlidwka) [inline lib/local-storage.js:35]: Private packages stored in the database with replication, public packages stored on the disk nobody cares about. This is an example of how people would like to use 2 storages at the same time.

**Comment by rlidwka** [comment]
@curtisr7 , I don't have much time nowadays, sorry for the delay.

Can you show an example plugin that could work with this? It's really the most important part for me to understand how it works.

And yes, I totally support any attempts to make storages pluggable, many people requested it, so we should implement something like this eventually.

File: lib/local-storage.js (hunk 2 above)
Comment (curtisr7) [inline lib/local-storage.js:35]: I still don't have head around all of the capabilities of what is/isn't possible. That makes sense though. Public modules are a cache, but the private published modules are records.

**Comment by curtisr7** [comment]
> Can you show an example plugin that could work with this? It's really the most important part for me to understand how it works.

Yes, I'll push something out shortly. I realize looking through the diffs my changes are all over the place... sorry about that.

A high level of what I've tried to do is factor all of the Store-centric operations into a single pluggable interface such that the higher level (local-store) is blissfully unaware of how/where data is being read/written. In my next commit I'll add a delegating store that does nothing more than send read/write operations to a pair of modules. As a first pass I wrote a module(couchdb-json-store) that reads/writes json to/fro a couchdb instance. Tomorrow I'm going to work on attaching tgz files to the couchdb record as attachments.

A few more comments/thoughts:
- We can improve perf a bunch, but I'm still in the poc phase. 
- Need to spend some time looking at LocalData to see how it could read/write through this same Store interface. 
- Search is very still very broken, but I haven't given it much thought.

There is an example configuration snippet in example-store-conf.yaml.

**Comment by curtisr7** [comment]
Doing some house cleaning. Closing this PR as I don't think I'm going to get back to this in the near term.

**Pull request closed without merging**